src/thunderbots/software/ai/hl/stp/tactic/shoot_goal_tactic.cpp

```diff
@@ -0,0 +1,161 @@
+        return std::make_pair(best_shot_opt->first, net_percent_open);
+    }
+
+    return std::nullopt;
+}
+
+void ShootGoalTactic::shootUntilShotBlocked(KickAction &kick_action,
+                                            ChipAction &chip_action,
+                                            IntentCoroutine::push_type &yield) const
+{
+    auto shot_target = getShotData();
+    while (shot_target)
+    {
+        if (!isEnemyAboutToStealBall())
+        {
+            // If we are in the middle of committing to a shot but an enemy is about to
+            // steal the ball we chip instead to just get over the enemy. We do not adjust
+            // the point we are targeting since that may take more time to realign to, and
+            // we need to be very quick so the enemy doesn't get the ball
+            yield(kick_action.updateStateAndGetNextIntent(
+                *robot, ball, ball.position(), shot_target->first,
+                BALL_MAX_SPEED_METERS_PER_SECOND - 0.5));
+        }
+        else
+        {
+            yield(chip_action.updateStateAndGetNextIntent(*robot, ball, ball.position(),
+                                                          shot_target->first, CHIP_DIST));
+        }
+
+        shot_target = getShotData();
+    }
+}
+
+void ShootGoalTactic::calculateNextIntent(IntentCoroutine::push_type &yield)
+{
+    KickAction kick_action = KickAction();
+    ChipAction chip_action = ChipAction();
+    MoveAction move_action = MoveAction(MoveAction::ROBOT_CLOSE_TO_DEST_THRESHOLD, true);
+
+    Point fallback_chip_target = chip_target ? *chip_target : field.enemyGoal();
+    do
+    {
+        auto shot_target = getShotData();
+        if (shot_target && shot_target->second > min_percent_net_open)
+        {
+            // Once we have determined we can take a shot, continue to try shoot until the
+            // shot is entirely blocked
+            has_shot_available = true;
+            shootUntilShotBlocked(kick_action, chip_action, yield);
+            has_shot_available = false;
+        }
+        else if (isEnemyAboutToStealBall())
+        {
+            // If an enemy is about to steal the ball from us, we try chip over them to
+            // try recover the ball after, which is better than being stripped of the ball
+            // and directly losing possession that way
+            yield(chip_action.updateStateAndGetNextIntent(*robot, ball, ball.position(),
+                                                          fallback_chip_target, 0.5));
+        }
+        else
+        {
+            Vector behind_ball_vector = (ball.position() - field.enemyGoal());
+            // A point behind the ball that leaves 5cm between the ball and kicker of the
+            // robot
+            Point behind_ball =
+                ball.position() +
+                behind_ball_vector.norm(BALL_MAX_RADIUS_METERS +
+                                        DIST_TO_FRONT_OF_ROBOT_METERS + 0.05);
```

Comment:
Can we make the `0.05` here a `const` member?

Reply:
done